ChatSpace データベース設計

yukitaka0915-se opened the pull request:
What
ChatSpaceを実際に触りながら、エンティティを洗い出してREADME.rmにデータベース設計を記述した。

Why
これから自身でChatSpaceを作る上で、データベース設計が最初の一歩となるため。

yukitaka0915-se committed:
add README.md

chat-space databese design  created

yukitaka0915-se committed:
add h1 title

TOMARUMARU commented:
多対多のアソシエーションについてよく調べて再度修正をお願いします

yukitaka0915-se committed:
Update README.md

fixed associations

Many-to-many：
messages, groups_users

One-to-many；
users, groups

yukitaka0915-se commented:
> 多対多のアソシエーションについてよく調べて再度修正をお願いします

多対多のアソシエーションを修正しました。
また、group_usersも中間テーブルになるので、
「groups_users」に改名しました。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+|Column|Type|Options|
+|------|----|-------|
+|id|integer|primary_key: true|
+|name|string|null: false|
```
usersテーブルのnameカラムはユーザー検索に使用するので検索を高速化するため、Optionsに`index: true`を追加するといいでしょう。

yukitaka0915-se replied:
nameカラムにindexを追加しました。

チャットグループを作成するときにユーザー検索をしていましたので、
仰るとおりですね。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+
+### Association
+- has_many :groups_users
+- has_many :groups, through: groups_users
```
細かいですが、38行目後半の`through: groups_users`に`:`を一つ足して`through: :groups_users`としましょう。

yukitaka0915-se replied:
修正しました。

38行目と51行目が該当していました。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+|------|----|-------|
+|id|integer|primary_key: true|
+|name|string|null: false|
+|user_id|integer|null: false, foreign_key: true|
```
groupsテーブルusersテーブルは多対多関係のため、47行目のuser_idカラムは不要ですので削除しましょう。

yukitaka0915-se replied:
修正しました。

多対多関係にあるテーブルでは、
中間テーブル（groups_usersテーブルとmessagesテーブル）が橋渡しをするので、
groupsテーブルとusersテーブルで互いのIDを持たなくて良いということですね。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+
+### Association
+- has_many :groups_users
+- has_many :users, through: groups_users
```
51行目後半の`through: groups_users`に`:`を一つ足して`through: :groups_users`としましょう。

yukitaka0915-se replied:
修正しました。

38行目の修正のコミットに含んでいます。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+|------|----|-------|
+|id|integer|primary_key: true|
+|group_id|integer|null: false, foreign_key: true|
+|user_id|integer|null: false, foreign_key: true|
```
59-60行目の外部キーはinteger型ではなく、references型を使用しましょう。
そうでない場合、`foreign_key: true`では外部キー制約にならないので注意です。
また、そうすることで `user_id` を `user` と記述でき、テーブルが生成される時に _id が付与されます。
例：`|user|references|null: false, foreign_key: true|`
参考：https://qiita.com/ryouzi/items/2682e7e8a86fd2b1ae47

yukitaka0915-se replied:
外部キーをreference型に、カラム名も_idを削除しました。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+|Column|Type|Options|
+|------|----|-------|
+|id|integer|primary_key: true|
+|body|string|null: false|
```
bodyとimageはどちらも単体で保存できるようにするため、現時点では`null: false`は削除しましょう。

yukitaka0915-se replied:
修正しました。

i-date commented on README.md:
```diff
@@ -22,3 +22,57 @@ Things you may want to cover:
+|body|string|null: false|
+|image|string||
+|group_id|integer|null: false, foreign_key: true|
+|user_id|integer|null: false, foreign_key: true|
```
73-74行目も外部キーですのでinteger型ではなく、references型を使用しましょう。

yukitaka0915-se replied:
修正しました。

こちらもreference型にすることで、カラム名から_idを削除しました。

i-date reviewed (commented):
ご提出ありがとうございます！コメントいたしました^^
また、すべての修正が完了したら再度レビューを依頼しましょう。

yukitaka0915-se committed:
Update README.md

add index for users
column name

yukitaka0915-se committed:
Update README.md

fixed
 through: groups_users
 through: :groups_users

yukitaka0915-se committed:
Update README.md

fixed groups table
delete column user_id

Not required because it is referenced through the groups_users table

yukitaka0915-se committed:
Update README.md

fixed groups_users table reference
 type_changed integer to reference
 rename to columnneme
  group_id → group
  user_id → user

yukitaka0915-se committed:
Update README.md

fixed options to [body]

Removed null: false so that body and image can be saved alone.

yukitaka0915-se committed:
Update README.md

fixed messages table reference
type_changed integer to reference
rename to columnneme
group_id → group
user_id → user

i-date reviewed (approved):
修正ありがとうございます！LGTMですのでマージしましょう^^
また、これから新機能の追加や新たにマークアップを行う際（特にカリキュラムでの指定箇所）は、忘れずに新しいブランチを切ってプルリクを作成しましょう。
